### ISSUE TITLE
Cell text missing first few letters - Too many rulings 

### ISSUE THREAD

**Issue by ghufda** [tabulapdf/tabula-java#166]
Hello !
I tried using the command  on the attached pdf on page 4.

java -jar "tabula-0.9.2-jar-with-dependencies.jar" "germony_1.pdf"  -r -g -f JSON  --pages 4

However the output has missing letters in the text. I also put the debug tool output and found that there are a lot of useless ruling detected.

[germony_1_page_4_table_2.txt](https://github.com/tabulapdf/tabula-java/files/1161823/germony_1_page_4_table_2.txt)

![germony_1-4_r](https://user-images.githubusercontent.com/30292326/28409867-36115aa8-6d3c-11e7-94f2-be8812cf2fae.jpg)
[germony_1.pdf](https://github.com/tabulapdf/tabula-java/files/1161771/germony_1.pdf)

Can someone give me some pointers on what part of the code is doing the ruling detection ( I am lost without comments ) or any ideas on how to improve on the existing heuristics ? 

Thanks,

**Comment by jazzido** [comment]
Hi @ghufda, thanks for reporting the issue.

Some PDFs, such as yours, contain "invisible" rulings around text. The solution for this would be to implement a filter for `Ruling`s in `ObjectExtractorStreamEngine`. We do have in fact a `TODO` in that class: https://github.com/tabulapdf/tabula-java/blob/master/src/main/java/technology/tabula/ObjectExtractorStreamEngine.java#L254

In [`tabula-extractor`](https://github.com/tabulapdf/tabula-extractor), the previous incarnation of `tabula-java`, we had such a feature: https://github.com/tabulapdf/tabula-extractor/blob/cf6d1982b94314b1effcc578547d178316ac4524/test/tests.rb#L369

Of course, we always welcome pull-requests :)

**Comment by criztovyl** [comment]
btw, have you tried not using stream (non-spreadsheet) mode?

`[tabula...] -g -f CSV --pages 4`

https://gist.github.com/criztovyl/7de1d0516e102e8099b63af50a7f97c8

**Comment by ghufda** [comment]
@criztovyl Yes I tried that but then cells are split into multiple parts, which is not ideal either.

@jazzido Thanks for the Links, I was finally able to find the piece of code that does the extraction but I am still trying to understand how it works. I will test the filter idea. 

Thanks a lot.

Youness

**Comment by ghufda** [comment]
Hello !
Do you know how I can access the color of a ruling and its backgroud ? I looked but could not figure it out by myself.
Thanks !
